File: PalindromeLinkedList.cpp
#include<iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        return isPalindrome(head, NULL);
    }
    
    bool isPalindrome(ListNode *begin, ListNode *end){
        if(begin == end || begin->next == end) return true;
        ListNode *p = begin;
        while(p->next != end) p = p->next;
        if(begin->val == p->val) return isPalindrome(begin->next, p);
        else return false;
    }
};

int main(){
    
    int n;
    cin>>n;
    
    ListNode *head = new ListNode(0);
    ListNode *p = head;
    for(int i=0;i<n;i++){
        int num;
        cin>>num;
        p->next = new ListNode(num);
        p = p->next;
    }
    head = head->next;
    
    Solution *solution = new Solution();
    solution->isPalindrome(head) ? cout<<"True" : cout<<"False";
    
    return 0;
    
}

File: PalindromeLinkedList.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        p = head
        cur = head
        pre = None
        while p and p.next:
            p = p.next.next
            next = cur.next
            cur.next = pre
            pre = cur
            cur = next
        if p:
            cur = cur.next

        while pre and cur:
            if pre.val != cur.val:
                return False
            pre = pre.next
            cur = cur.next
        return True

if __name__ == '__main__':
    str_in = input()
    nums = [int(n) for n in str_in.split()]
    head = ListNode(0)
    p = head
    for num in nums:
        p.next = ListNode(num)
        p = p.next
    head = head.next
    solution = Solution()
    print(solution.isPalindrome(head))